# Pandas
- useful for data processing and analysis
- has two objects:
  - pandas dataframe
  - pandas series

## Pandas DataFrame
- two dimensionaltabluar data structure with labeled axis (rows and columns)

In [213]:
from google.colab import files
dataset = files.upload()


Saving diabetes.csv to diabetes (6).csv


In [214]:
import pandas as pd

In [215]:
from sklearn.datasets import fetch_california_housing
housing = fetch_california_housing()
print(type(housing))

<class 'sklearn.utils._bunch.Bunch'>


In [216]:
housing

{'data': array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
           37.88      , -122.23      ],
        [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
           37.86      , -122.22      ],
        [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
           37.85      , -122.24      ],
        ...,
        [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
           39.43      , -121.22      ],
        [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
           39.43      , -121.32      ],
        [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
           39.37      , -121.24      ]]),
 'target': array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894]),
 'frame': None,
 'target_names': ['MedHouseVal'],
 'feature_names': ['MedInc',
  'HouseAge',
  'AveRooms',
  'AveBedrms',
  'Population',
  'AveOccup',
  'Latitude',
  'Longitude'],
 'DESCR': '.. _california_housing_dataset:\n

- The above dataset is in the form of key value pair. This is difficult to process adn analyse. Hence DataFrame comes into picture

# Converting to a pandas DataFrame

In [217]:
housing_df = pd.DataFrame(housing.data, columns = housing.feature_names)

In [218]:
housing_df.head(5)

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

In [219]:
housing_df.shape

(20640, 8)

# Reading a csv file

In [220]:
diabetes_df = pd.read_csv('/content/diabetes.csv')
print(type(diabetes_df))

<class 'pandas.core.frame.DataFrame'>


In [221]:
diabetes_df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [222]:
diabetes_df.shape

(768, 9)

# loading an excel file as dataframe
- code: excelFile = pd.read_excel('file_path')

# Converting a dataframe to csv

In [223]:
housing_df.to_csv('housing.csv')

# Converting df to excel:
- df.to_excel('name.xlsx')

# Creating a dataframe with random values

In [224]:
import numpy as np
random_df = pd.DataFrame(np.random.rand(20,10)) #values between 0 & 1
random_df.head()

0         1         2         3         4         5         6  \
0  0.534937  0.264350  0.435220  0.840053  0.650144  0.500078  0.040642   
1  0.861333  0.447017  0.018870  0.178097  0.338728  0.277074  0.000786   
2  0.506481  0.055055  0.343938  0.863350  0.774438  0.259759  0.487254   
3  0.691331  0.739498  0.109188  0.311065  0.377721  0.880066  0.438140   
4  0.370120  0.335704  0.383423  0.748313  0.917035  0.479731  0.747476   

          7         8         9  
0  0.601701  0.532404  0.824216  
1  0.203976  0.098468  0.128517  
2  0.567545  0.776952  0.355989  
3  0.488061  0.103516  0.567101  
4  0.750616  0.947214  0.595969

# Inspecting a datset

In [225]:
housing_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

In [226]:
housing_df.shape

(20640, 8)

In [227]:
#last 5 rows
housing_df.tail()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  
20635    -121.09  
20636    -121.21  
20637    -121.22  
20638    -121.32  
20639    -121.24

In [228]:
housing_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   MedInc      20640 non-null  float64
 1   HouseAge    20640 non-null  float64
 2   AveRooms    20640 non-null  float64
 3   AveBedrms   20640 non-null  float64
 4   Population  20640 non-null  float64
 5   AveOccup    20640 non-null  float64
 6   Latitude    20640 non-null  float64
 7   Longitude   20640 non-null  float64
dtypes: float64(8)
memory usage: 1.3 MB


In [229]:
#no of missong values
housing_df.isnull().sum()

MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
dtype: int64

In [230]:
diabetes_df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [231]:
# Counting no of diabetic and non-diabetic patients
diabetes_df.value_counts('Outcome')

Outcome
0    500
1    268
Name: count, dtype: int64

In [232]:
# finding mean of every column for diabetic and non-diabetic
diabetes_df.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

## Statistical measures

In [233]:
# counting no of values in each column
housing_df.count()

MedInc        20640
HouseAge      20640
AveRooms      20640
AveBedrms     20640
Population    20640
AveOccup      20640
Latitude      20640
Longitude     20640
dtype: int64

In [234]:
# mean of values in each column
housing_df.mean()

MedInc           3.870671
HouseAge        28.639486
AveRooms         5.429000
AveBedrms        1.096675
Population    1425.476744
AveOccup         3.070655
Latitude        35.631861
Longitude     -119.569704
dtype: float64

In [235]:
# max of values in each column
housing_df.max()

MedInc           15.000100
HouseAge         52.000000
AveRooms        141.909091
AveBedrms        34.066667
Population    35682.000000
AveOccup       1243.333333
Latitude         41.950000
Longitude      -114.310000
dtype: float64

In [236]:
# min of values in each column
housing_df.min()

MedInc          0.499900
HouseAge        1.000000
AveRooms        0.846154
AveBedrms       0.333333
Population      3.000000
AveOccup        0.692308
Latitude       32.540000
Longitude    -124.350000
dtype: float64

In [237]:
# standard deviation of values in each column
housing_df.std()

MedInc           1.899822
HouseAge        12.585558
AveRooms         2.474173
AveBedrms        0.473911
Population    1132.462122
AveOccup        10.386050
Latitude         2.135952
Longitude        2.003532
dtype: float64

In [238]:
#all statistical measures at once
housing_df.describe()

MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude  
count  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704  
std       10.386050      2.135952      2.003532  
min        0.692308     32.540000   -124.350000  
25%        2.429741     33.930000   -121.800000  
50%        2.818116     34.260000   -118.490000  
75%        3.282261     37.710000   -118.010000  
max     1243.333333     41.950000   -114.310000

# Manipulating the dataframe

The housing dataset has a feature called target but it is not included in dataframe

## adding column in dataframe
- Note: to this no of data points must be same.
- The og housing dataset had 20640 data points. Hence, the target column must also have 20640 data points


In [239]:
housing_df.shape

(20640, 8)

In [240]:
# creating a new column called price

housing_df['Price'] = housing.target
housing_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  Price  
0    -122.23  4.526  
1    -122.22  3.585  
2    -122.24  3.521  
3    -122.25  3.413  
4    -122.25  3.422

## Removing rows and columns

In [241]:
housing_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  Price  
0    -122.23  4.526  
1    -122.22  3.585  
2    -122.24  3.521  
3    -122.25  3.413  
4    -122.25  3.422

In [242]:
#removing row (give index i.e. which row number and axis = 0 indicates row removal)
housing_df.drop(index=1,axis=0,inplace = True)
housing_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
5  4.0368      52.0  4.761658   1.103627       413.0  2.139896     37.85   

   Longitude  Price  
0    -122.23  4.526  
2    -122.24  3.521  
3    -122.25  3.413  
4    -122.25  3.422  
5    -122.25  2.697

In [243]:
# removing column. Give column name in and axis = 1 indicates column removal
housing_df.drop(columns='Population',axis=1,inplace = True)
housing_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  AveOccup  Latitude  Longitude  Price
0  8.3252      41.0  6.984127   1.023810  2.555556     37.88    -122.23  4.526
2  7.2574      52.0  8.288136   1.073446  2.802260     37.85    -122.24  3.521
3  5.6431      52.0  5.817352   1.073059  2.547945     37.85    -122.25  3.413
4  3.8462      52.0  6.281853   1.081081  2.181467     37.85    -122.25  3.422
5  4.0368      52.0  4.761658   1.103627  2.139896     37.85    -122.25  2.697

## Loacting or accessing a particular row using index value

In [244]:
housing_df.iloc[2] # gives each column value for row indexed 2

MedInc         5.643100
HouseAge      52.000000
AveRooms       5.817352
AveBedrms      1.073059
AveOccup       2.547945
Latitude      37.850000
Longitude   -122.250000
Price          3.413000
Name: 3, dtype: float64

In [247]:
print(housing_df.iloc[:,0]) # first column
print()
print(housing_df.iloc[:,1]) # second column
print()
print(housing_df.iloc[:,2]) # third column
print()
print(housing_df.iloc[:,-1]) # last column

0        8.3252
2        7.2574
3        5.6431
4        3.8462
5        4.0368
          ...  
20635    1.5603
20636    2.5568
20637    1.7000
20638    1.8672
20639    2.3886
Name: MedInc, Length: 20639, dtype: float64

0        41.0
2        52.0
3        52.0
4        52.0
5        52.0
         ... 
20635    25.0
20636    18.0
20637    17.0
20638    18.0
20639    16.0
Name: HouseAge, Length: 20639, dtype: float64

0        6.984127
2        8.288136
3        5.817352
4        6.281853
5        4.761658
           ...   
20635    5.045455
20636    6.114035
20637    5.205543
20638    5.329513
20639    5.254717
Name: AveRooms, Length: 20639, dtype: float64

0        4.526
2        3.521
3        3.413
4        3.422
5        2.697
         ...  
20635    0.781
20636    0.771
20637    0.923
20638    0.847
20639    0.894
Name: Price, Length: 20639, dtype: float64


# Correlation
1. positive correlation -> one variable increases then other also increases
2. negative correlation -> one variable increases then other decreases

In [248]:
housing_df.corr()

MedInc  HouseAge  AveRooms  AveBedrms  AveOccup  Latitude  \
MedInc     1.000000 -0.118982  0.326902  -0.062019  0.018779 -0.079940   
HouseAge  -0.118982  1.000000 -0.153270  -0.077756  0.013189  0.011204   
AveRooms   0.326902 -0.153270  1.000000   0.847629 -0.004851  0.106376   
AveBedrms -0.062019 -0.077756  0.847629   1.000000 -0.006182  0.069736   
AveOccup   0.018779  0.013189 -0.004851  -0.006182  1.000000  0.002371   
Latitude  -0.079940  0.011204  0.106376   0.069736  0.002371  1.000000   
Longitude -0.015029 -0.108241 -0.027520   0.013328  0.002470 -0.924661   
Price      0.688046  0.105667  0.151934  -0.046686 -0.023733 -0.144237   

           Longitude     Price  
MedInc     -0.015029  0.688046  
HouseAge   -0.108241  0.105667  
AveRooms   -0.027520  0.151934  
AveBedrms   0.013328 -0.046686  
AveOccup    0.002470 -0.023733  
Latitude   -0.924661 -0.144237  
Longitude   1.000000 -0.045886  
Price      -0.045886  1.000000